Add stack_trace on make_fx

yushangdi opened the pull request:
Summary:
Previosuly, we only add stack trace in `class _ModuleStackTracer(PythonKeyTracer)` for non-strict export. I moved this stack trace logic to the parent class `PythonKeyTracer`, this way the graph traced from Module using make_fx will have stack_trace as well.

Motivation: we've observed some uses cases where users first use `make_fx` on the Module, and then run `export` on the resulting graph. If the result of `make_fx` doesn't have stack trace, the stack trace information is lost.

Test Plan:
```
buck run test:test_export -- -r  test_stack_trace
```

Rollback Plan:

Differential Revision: D75985427




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

angelayi reviewed (approved):
paraphrased from @zou3519: can we check that the stacktrace added here doesn't affect torch.compile node stacktraces, where dynamo generates the stacktrace, and then this codepath gets called in AOTAutograd?

pretty sure this code doesn't affect the torch.compile stacktraces because of the `if "stack_trace" not in node.meta` conditional, but let's just check to make sure

yushangdi commented:
> paraphrased from @zou3519: can we check that the stacktrace added here doesn't affect torch.compile node stacktraces, where dynamo generates the stacktrace, and then this codepath gets called in AOTAutograd?
> 
> pretty sure this code doesn't affect the torch.compile stacktraces because of the `if "stack_trace" not in node.meta` conditional, but let's just check to make sure

Yep it's still there!

```
        inp = torch.randn(4, 4)

        graph = None
        def my_compiler(gm: torch.fx.GraphModule, example_inputs: list[torch.Tensor]):
            global graph
            graph = gm.graph
            gm.print_readable()
            return gm.forward  # return a python callable

        result = torch.compile(Foo(), backend=my_compiler)(inp)
   ```
   
   ```
class GraphModule(torch.nn.Module):
    def forward(self, L_self_modules_linear_parameters_weight_: "f32[4, 4]", L_self_modules_linear_parameters_bias_: "f32[4]", L_x_: "f32[4, 4]"):
        l_self_modules_linear_parameters_weight_ = L_self_modules_linear_parameters_weight_
        l_self_modules_linear_parameters_bias_ = L_self_modules_linear_parameters_bias_
        l_x_ = L_x_
        
         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/a7ab72042f458866/caffe2/test/inductor/__provenance_tracing__/provenance_tracing#link-tree/caffe2/test/inductor/test_provenance_tracing.py:417 in forward, code: x = self.linear(x)
        x: "f32[4, 4]" = torch._C._nn.linear(l_x_, l_self_modules_linear_parameters_weight_, l_self_modules_linear_parameters_bias_);  l_x_ = l_self_modules_linear_parameters_weight_ = l_self_modules_linear_parameters_bias_ = None
        
         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/a7ab72042f458866/caffe2/test/inductor/__provenance_tracing__/provenance_tracing#link-tree/caffe2/test/inductor/test_provenance_tracing.py:418 in forward, code: x *= 2.0
        x *= 2.0;  x_1: "f32[4, 4]" = x;  x = None
        return (x_1,)
   
   
   ```

yushangdi committed:
Add stack_trace on make_fx (#155155)

Summary:

Previosuly, we only add stack trace in `class _ModuleStackTracer(PythonKeyTracer)` for non-strict export. I moved this stack trace logic to the parent class `PythonKeyTracer`, this way the graph traced from Module using make_fx will have stack_trace as well.

Motivation: we've observed some uses cases where users first use `make_fx` on the Module, and then run `export` on the resulting graph. If the result of `make_fx` doesn't have stack trace, the stack trace information is lost.

Test Plan:
```
buck run test:test_export -- -r  test_stack_trace
```

Rollback Plan:

Reviewed By: angelayi

Differential Revision: D75985427

zou3519 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1110,6 +1110,35 @@ def create_node(
         node = super().create_node(kind, target, args, kwargs, name, type_expr)  # type: ignore[arg-type]
 
+        # stack_trace
+        if "stack_trace" not in node.meta and node.op not in ["placeholder", "output"]:
```
When would stack_trace already be in node.meta? This makes it a bit difficult to reason about this

zou3519 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1110,6 +1110,35 @@ def create_node(
+                stack_trace = [
+                    frame
+                    for frame in user_frame_summary
+                    if (
+                        frame.name == "forward"
+                        or frame.filename.endswith("torch/__init__.py")
+                    )
```
Does similar logic exist somewhere else already? I can just write a function named `forward` and it would not retain the stack trace. It sounded like you wanted a more robust check for nn.Module.forward

yushangdi commented:
close. will try to reland in https://github.com/pytorch/pytorch/pull/155486
